File: loadercomplex-forge/ver12/src/main/java/com/enderzombi102/loadercomplex/forge12/impl/utils/ForgeRegistry.java
package com.enderzombi102.loadercomplex.forge12.impl.utils;

import com.enderzombi102.loadercomplex.api.utils.Registry;
import com.enderzombi102.loadercomplex.minecraft.block.Block;
import com.enderzombi102.loadercomplex.minecraft.entity.Entity;
import com.enderzombi102.loadercomplex.minecraft.item.Item;
import com.enderzombi102.loadercomplex.minecraft.util.RegistryKey;
import com.enderzombi102.loadercomplex.minecraft.util.ResourceIdentifier;
import com.enderzombi102.loadercomplex.forge12.LoaderComplexForge;
import com.enderzombi102.loadercomplex.forge12.impl.block.ForgeBlock;
import com.enderzombi102.loadercomplex.forge12.impl.item.ForgeItem;
import net.minecraft.client.renderer.block.model.ModelResourceLocation;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraftforge.client.model.ModelLoader;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import net.minecraftforge.registries.GameData;
import net.minecraftforge.registries.RegistryManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import static com.enderzombi102.loadercomplex.minecraft.util.ResourceIdentifier.ri;

public class ForgeRegistry implements Registry {
	private final HashMap<ResourceIdentifier, net.minecraft.block.Block> blocks = new HashMap<>();
	private final HashMap<ResourceIdentifier, net.minecraft.item.Item> items = new HashMap<>();
	private static final Map<ResourceIdentifier, CreativeTabs> ITEM_GROUPS = new HashMap<ResourceIdentifier, CreativeTabs>() {{
		put( ri( "itemgroup.brewing" ), CreativeTabs.BREWING );
		put( ri( "itemgroup.building_blocks" ), CreativeTabs.BUILDING_BLOCKS );
		put( ri( "itemgroup.combat" ), CreativeTabs.COMBAT );
		put( ri( "itemgroup.decorations" ), CreativeTabs.DECORATIONS );
		put( ri( "itemgroup.food" ), CreativeTabs.FOOD );
		put( ri( "itemgroup.materials" ), CreativeTabs.MATERIALS );
		put( ri( "itemgroup.redstone" ), CreativeTabs.REDSTONE );
		put( ri( "itemgroup.tools" ), CreativeTabs.TOOLS );
		put( ri( "itemgroup.transportation" ), CreativeTabs.TRANSPORTATION );
		put( ri( "itemgroup.misc" ), CreativeTabs.MISC );
	}};

	@Override
	public void register(@NotNull Block block, @NotNull ResourceIdentifier identifier) {
		this.register( block, identifier, ri( "itemgroup.misc" ) );
	}

	@Override
	public void register(@NotNull Block block, @NotNull ResourceIdentifier identifier, @Nullable ResourceIdentifier itemGroup ) {
		final ForgeBlock forgeBlock = new ForgeBlock( block );
		forgeBlock.setUnlocalizedName( identifier.getNamespace() + '.' + identifier.getPath() );
		this.blocks.put(identifier, forgeBlock);
		if ( itemGroup != null )
			this.items.put(
				identifier,
				new ItemBlock( forgeBlock )
					.setUnlocalizedName( identifier.getNamespace() + '.' + identifier.getPath() )
					.setCreativeTab( getOrCreateItemGroup( itemGroup, identifier ) )
			);
	}

	@Override
	public void register(@NotNull Item item, @NotNull ResourceIdentifier identifier) {
		this.items.put(
			identifier,
			new ForgeItem( item )
				.setUnlocalizedName( identifier.getNamespace() + '.' + identifier.getPath() )
				.setCreativeTab( getOrCreateItemGroup( item.group, identifier ) )
		);
	}

	@Override
	public void register(@NotNull Entity entity, @NotNull ResourceIdentifier identifier) {

	}

	@Override
	public ResourceIdentifier registerItemGroup( @Nullable String name, @NotNull ResourceIdentifier icon ) {
		ResourceIdentifier id = new ResourceIdentifier( icon.getNamespace(), name != null ? name : icon.getNamespace() );
		ITEM_GROUPS.computeIfAbsent(
			id,
			key -> new CreativeTabs( name != null ? icon.getNamespace() + "." + name : icon.getNamespace() ) {
				@SideOnly(Side.CLIENT)
				public @NotNull ItemStack getTabIconItem() {
					return new ItemStack(
						Objects.requireNonNull(
							net.minecraft.item.Item.getByNameOrId( icon.toString() ),
							String.format( "The icon given by %s was invalid", icon.getNamespace() )
						)
					);
				}
			}
		);
		return id;
	}

	@Override
	public boolean isRegistered(@NotNull RegistryKey key, @NotNull String identifier) {
		return false;
	}

	@Override
	public boolean isRegistered(@NotNull RegistryKey key, @NotNull ResourceIdentifier identifier) {
		return false;
	}

	@Override
	public Object getVanillaRegistry(@NotNull RegistryKey type) {
		switch (type) {
			case Item:
				return RegistryManager.ACTIVE.getRegistry(GameData.ITEMS);
			case Block:
				return RegistryManager.ACTIVE.getRegistry(GameData.BLOCKS);
			default:
				return null;
		}
	}

	public void onItemRegistry(RegistryEvent.Register<net.minecraft.item.Item> evt) {
		this.items.forEach( ( id, item ) -> evt.getRegistry().register( item.setRegistryName( id.toString() ) ) );
	}

	public void onBlockRegistry(RegistryEvent.Register<net.minecraft.block.Block> evt) {
		this.blocks.forEach( ( id, block ) -> evt.getRegistry().register( block.setRegistryName( id.toString() ) ) );
	}

	public void onModelRegistry() {
		this.items.forEach( ( id, item ) -> ModelLoader.setCustomModelResourceLocation(
			item,
			0,
			new ModelResourceLocation( id.toString(), "inventory" )
		));
	}

	public static CreativeTabs getOrCreateItemGroup(@Nullable ResourceIdentifier itemGroup, ResourceIdentifier icon ) {
		if ( itemGroup == null )
			return null;
		if (! ITEM_GROUPS.containsKey( itemGroup ) )
			LoaderComplexForge.INSTANCE.getLoader().getRegistry().registerItemGroup( null, icon );
		return ITEM_GROUPS.get( itemGroup );
	}
}

File: loadercomplex-forge/src/ver122/kotlin/com/enderzombi102/loadercomplex/forge122/impl/platform/ForgeRegistry.kt
package com.enderzombi102.loadercomplex.forge122.impl.platform

import com.enderzombi102.loadercomplex.api.minecraft.block.Block
import com.enderzombi102.loadercomplex.api.minecraft.entity.Entity
import com.enderzombi102.loadercomplex.api.minecraft.item.Item
import com.enderzombi102.loadercomplex.api.minecraft.util.RegistryKey
import com.enderzombi102.loadercomplex.api.minecraft.util.ResourceIdentifier
import com.enderzombi102.loadercomplex.api.minecraft.util.ResourceIdentifier.ri
import com.enderzombi102.loadercomplex.api.platform.Registry
import com.enderzombi102.loadercomplex.forge122.impl.block.ForgeBlock
import com.enderzombi102.loadercomplex.forge122.impl.item.ForgeItem
import com.enderzombi102.loadercomplex.impl.LoaderComplex
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.ItemBlock
import net.minecraft.item.ItemStack
import net.minecraftforge.client.model.ModelLoader
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.minecraftforge.registries.GameData
import net.minecraftforge.registries.RegistryManager
import java.util.*
import net.minecraft.block.Block as McBlock
import net.minecraft.item.Item as McItem

class ForgeRegistry : Registry {
	private val blocks = HashMap<ResourceIdentifier, McBlock>()
	private val items = HashMap<ResourceIdentifier, McItem>()
	override fun register( block: Block, identifier: ResourceIdentifier ) {
		this.register( block, identifier, ri( "itemgroup.misc" ) )
	}

	override fun register( block: Block, identifier: ResourceIdentifier, itemGroup: ResourceIdentifier? ) {
		val forgeBlock = ForgeBlock( block )
		forgeBlock.setUnlocalizedName( identifier.namespace + '.' + identifier.path )
		blocks[identifier] = forgeBlock
		if ( itemGroup != null ) {
			items[identifier] = ItemBlock( forgeBlock )
				.setUnlocalizedName( identifier.namespace + '.' + identifier.path )
				.setCreativeTab( getOrCreateCreativeTab( itemGroup, identifier ) )
		}
	}

	override fun register( item: Item, identifier: ResourceIdentifier ) {
		items[identifier] = ForgeItem( item )
			.setUnlocalizedName( identifier.namespace + '.' + identifier.path )
			.setCreativeTab( getOrCreateCreativeTab( item.creativeTab, identifier ) )
	}

	override fun register( entity: Entity, identifier: ResourceIdentifier ) {
	}

	override fun registerCreativeTab( name: String?, icon: ResourceIdentifier ): ResourceIdentifier {
		val id = ResourceIdentifier( icon.namespace, name ?: icon.namespace )
		CREATIVE_TABS.computeIfAbsent( id ) {
			object : CreativeTabs( if ( name != null ) icon.namespace + "." + name else icon.namespace ) {
				@SideOnly( Side.CLIENT )
				override fun getTabIconItem(): ItemStack =
					ItemStack( requireNotNull( McItem.getByNameOrId( icon.toString() ) ) { String.format( "The icon given by %s was invalid", icon.namespace ) } )
			}
		}
		return id
	}

	override fun isRegistered( key: RegistryKey, identifier: String ): Boolean =
		false

	override fun isRegistered( key: RegistryKey, identifier: ResourceIdentifier ): Boolean =
		false

	override fun getVanillaRegistry( type: RegistryKey ): Any? = when (type) {
		RegistryKey.Item -> RegistryManager.ACTIVE.getRegistry<McItem>( GameData.ITEMS )
		RegistryKey.Block -> RegistryManager.ACTIVE.getRegistry<McBlock>( GameData.BLOCKS )
		else -> null
	}

	fun onItemRegistry( evt: RegistryEvent.Register<McItem> ) {
		items.forEach { ( id: ResourceIdentifier, item: McItem ) ->
			evt.registry.register( item.setRegistryName( id.toString() ) )
		}
	}

	fun onBlockRegistry( evt: RegistryEvent.Register<McBlock> ) {
		blocks.forEach { ( id: ResourceIdentifier, block: McBlock ) ->
			evt.registry.register( block.setRegistryName( id.toString() ) )
		}
	}

	fun onModelRegistry() {
		items.forEach { ( id: ResourceIdentifier, item: McItem ) ->
			ModelLoader.setCustomModelResourceLocation(
				item,
				0,
				ModelResourceLocation( id.toString(), "inventory" )
			)
		}
	}

	companion object {
		private val CREATIVE_TABS: MutableMap<ResourceIdentifier, CreativeTabs> =
			HashMap<ResourceIdentifier, CreativeTabs>().apply {
				put( ri( "creativetab.brewing" ), CreativeTabs.BREWING )
				put( ri( "creativetab.building_blocks" ), CreativeTabs.BUILDING_BLOCKS )
				put( ri( "creativetab.combat" ), CreativeTabs.COMBAT )
				put( ri( "creativetab.decorations" ), CreativeTabs.DECORATIONS )
				put( ri( "creativetab.food" ), CreativeTabs.FOOD )
				put( ri( "creativetab.materials" ), CreativeTabs.MATERIALS )
				put( ri( "creativetab.redstone" ), CreativeTabs.REDSTONE )
				put( ri( "creativetab.tools" ), CreativeTabs.TOOLS )
				put( ri( "creativetab.transportation" ), CreativeTabs.TRANSPORTATION )
				put( ri( "creativetab.misc" ), CreativeTabs.MISC )
			}

		fun getOrCreateCreativeTab( itemGroup: ResourceIdentifier?, icon: ResourceIdentifier ): CreativeTabs? {
			if ( itemGroup == null ) {
				return null
			}
			if (! CREATIVE_TABS.containsKey( itemGroup ) ) {
				LoaderComplex.get().context.registry.registerCreativeTab( null, icon )
			}
			return CREATIVE_TABS[itemGroup]
		}
	}
}
